### PR TITLE
Specs with sample app

### PR DESCRIPTION
**Pull request by hayesr**
This is the more official way of testing an engine, is a little more isolated, easier to setup with Travis, and follows the pattern of other ManageIQ splits.

There are [many spec failures](https://travis-ci.org/hayesr/manageiq-ui-classic/builds/184705611), but they seem to be mostly due to spec files being out of sync with the code they test. For example, [the quadicon helper](https://github.com/martinpovolny/manageiq-ui-classic/blob/ui_split_6/spec/helpers/quadicon_helper_spec.rb#L76). See also [this build](https://travis-ci.org/hayesr/manageiq-ui-classic/builds/184704033) (with [corresponding commit](https://github.com/hayesr/manageiq-ui-classic/commit/ea6267bc8b1d3b4b4aac454faa727fa7e51d1dfe)) as proof that this setup works.

This `spec_helper` includes a good deal of duplication from the core repo. A future todo would be to minimize that. See also https://github.com/ManageIQ/manageiq-content/issues/6

### CODE REVIEW

**Comment by hayesr** [comment]
Related: https://github.com/martinpovolny/manageiq-ui-classic/pull/1

**Comment by hayesr** [comment]
Related: https://github.com/ManageIQ/manageiq/issues/13215

File: Gemfile [hunk 1]
```diff
@@ -10,7 +10,23 @@ gemspec
...
+end
+
+group :test do
+  gem "codeclimate-test-reporter", :require => false
+  gem "simplecov", :require => false
+end
+
+group :development, :test do
+  gem "pry-rails"
+end
```
Comment (Fryguy) [inline Gemfile]: We do not have pry in our application, so we shouldn't add it here (I realize it's in @martinpovolny's branhc, but it doesn't belong there either).  If we all agree we want it, it should be discussed in the main ManageIQ repo as a separate event.

File: lib/manageiq/ui/engine.rb [hunk 2]
```diff
@@ -1,3 +1,4 @@
+require 'rails/all'
 require 'patternfly-sass'
```
Comment (Fryguy) [inline lib/manageiq/ui/engine.rb]: Suprised patternfly-sass needs such an early require.  @hayesr I noticed you removed patternfly-sass from the Gemfile.  That should probably be put back as I assume it was there for a reason.  On that note, if you have it in the Gemfile without a `:require => false`, then this line should not be necessary.

File: spec/spec_helper.rb [hunk 3]
```diff
@@ -0,0 +1,99 @@
+# require 'simplecov'
+# SimpleCov.start
+
+# TODO: Make this spec_helper stand alone.
+
+ENV["RAILS_ENV"] ||= 'test'
```
Comment (Fryguy) [inline spec/spec_helper.rb]: @bdunne Please take a look at this spec_helper.rb.  As we consolidate, i'm sure there are things here we'll want to share.

File: spec/spec_helper.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+require File.expand_path("../manageiq/config/environment", __FILE__)
+require 'application_helper'
+require 'rails-controller-testing'
+require 'rspec/rails'
+
+require 'manageiq/ui'
```
Comment (Fryguy) [inline spec/spec_helper.rb]: I would expect `require 'manageiq-ui-classic'`

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+RSpec.configure do |config|
+  config.expect_with :rspec do |c|
+    c.syntax = :expect
+  end
+  config.mock_with :rspec do |c|
+    c.syntax = :expect
+  end
+
+  config.fixture_path = "#{::Rails.root}/spec/fixtures"
```
Comment (Fryguy) [inline spec/spec_helper.rb]: `Rails.root.join("spec/fixtures")`

File: spec/spec_helper.rb [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+  config.infer_spec_type_from_file_location!
+
+  unless ENV['CI']
+    # File store for --only-failures option
+    config.example_status_persistence_file_path = Rails.root.join("tmp/rspec_example_store.txt")
+  end
+
+  config.define_derived_metadata(:file_path => /spec\/lib\/miq_automation_engine\/models/) do |metadata|
+    metadata[:type] ||= :model
```
Comment (Fryguy) [inline spec/spec_helper.rb]: I don't believe these are models.  Also, non of this code should be in this repo, so I don't think this whole block should be necessary.

File: spec/spec_helper.rb [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+  config.include Spec::Support::RakeTaskExampleGroup, :type => :rake_task
+  config.include Spec::Support::ButtonHelper, :type => :button
+  config.include Spec::Support::AuthHelper, :type => :button
+  config.define_derived_metadata(:file_path => /spec\/helpers\/application_helper\/buttons/) do |metadata|
+    metadata[:type] = :button
+  end
+
+  config.before(:each) do |example|
+    EmsRefresh.try(:debug_failures=, true) if example.metadata[:migrations].blank?
```
Comment (Fryguy) [inline spec/spec_helper.rb]: This should no be neccesary as EmsRefresh specs will not be in this app.

File: spec/spec_helper.rb [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+  config.include UiConstants, :type => :controller
+
+  config.include Spec::Support::AuthHelper, :type => :helper
+
+  config.include Spec::Support::PresenterHelper, :type => :presenter
+  config.define_derived_metadata(:file_path => /spec\/presenters/) do |metadata|
+    metadata[:type] ||= :presenter
+  end
+
+  config.include Spec::Support::RakeTaskExampleGroup, :type => :rake_task
```
Comment (Fryguy) [inline spec/spec_helper.rb]: I don't think there are rake test tasks in this app, so this can be removed.

File: spec/spec_helper.rb [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  config.define_derived_metadata(:file_path => /spec\/helpers\/application_helper\/buttons/) do |metadata|
+    metadata[:type] = :button
+  end
+
+  config.before(:each) do |example|
+    EmsRefresh.try(:debug_failures=, true) if example.metadata[:migrations].blank?
+    ApplicationController.handle_exceptions = false if %w(controller requests).include?(example.metadata[:type])
+  end
+
+  # config.before(:each, :rest_api => true) { init_api_spec_env }
```
Comment (Fryguy) [inline spec/spec_helper.rb]: Yeah, this can be removed...no need to comment.

File: spec/spec_helper.rb [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+Dir[ManageIQ::Gems::Pending.root.join("spec/support/custom_matchers/*.rb")].each { |f| require f }
+
+RSpec.configure do |config|
+  config.expect_with :rspec do |c|
+    c.syntax = :expect
+  end
+  config.mock_with :rspec do |c|
+    c.syntax = :expect
+  end
```
Comment (Fryguy) [inline spec/spec_helper.rb]: I think we can remove these 6 lines.  We haven't been putting them in the other repos.  Is Rspec basically expect syntax only at this point?  cc @chrisarcand ... should we have this or remove it?

File: spec/spec_helper.rb [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+Dir[support_path.join("**","*.rb")].each { |f| require f }
+
+shared_path = Rails.root.join('spec', 'shared', 'controllers', '**', '*.rb')
+
+Dir[shared_path].each do |f|
+  require f
+end
+
+require support_path.join('examples_group', 'shared_examples_for_application_helper.rb').to_s
+require support_path.join('rake_task_example_group.rb').to_s
```
Comment (Fryguy) [inline spec/spec_helper.rb]: This should not be necessary as we are already loading them on line 29

File: spec/spec_helper.rb [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+# require support_path.join('button_helper.rb')
+# require support_path.join('settings_helper.rb')
+
+Dir[support_path.join("**","*.rb")].each { |f| require f }
+
+shared_path = Rails.root.join('spec', 'shared', 'controllers', '**', '*.rb')
+
+Dir[shared_path].each do |f|
+  require f
+end
```
Comment (Fryguy) [inline spec/spec_helper.rb]: Minor, but can you clean this up to look like line 29?  They seem similar, so probably should look similar.

Additionally, it's really weird for shared to be at the root of the spec directory.  Was it already like that?  In the main repo we have spec/support/shared_examples.

File: spec/spec_helper.rb [hunk 13]
```diff
@@ -0,0 +1,99 @@
+# require 'simplecov'
+# SimpleCov.start
+
+# TODO: Make this spec_helper stand alone.
+
+ENV["RAILS_ENV"] ||= 'test'
+require File.expand_path("../manageiq/config/environment", __FILE__)
+require 'application_helper'
+require 'rails-controller-testing'
```
Comment (Fryguy) [inline spec/spec_helper.rb]: This gem should come over to this repo.

File: spec/spec_helper.rb [hunk 14]
```diff
@@ -0,0 +1,99 @@
+# require 'simplecov'
+# SimpleCov.start
+
+# TODO: Make this spec_helper stand alone.
+
+ENV["RAILS_ENV"] ||= 'test'
+require File.expand_path("../manageiq/config/environment", __FILE__)
+require 'application_helper'
```
Comment (Fryguy) [inline spec/spec_helper.rb]: Very strange that we are requiring ApplicationHelper...is this necessary?

**Comment by Fryguy** [comment]
Looking good @hayesr !

File: .rspec [hunk 15]
```diff
@@ -0,0 +1,3 @@
+--require spec_helper
```
Comment (chrisarcand) [inline .rspec]: Did you ever test if the line that loads MIQ's helper (copied from `content`) actually does execute both blocks? Looks like you've made it work without 👍

File: lib/manageiq/ui/engine.rb (hunk 2 above)
Comment (hayesr) [inline lib/manageiq/ui/engine.rb]: It gives an error about being in both gemfiles (until we remove from the main one)

File: spec/spec_helper.rb [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+  unless ENV['CI']
+    # File store for --only-failures option
+    config.example_status_persistence_file_path = Rails.root.join("tmp/rspec_example_store.txt")
+  end
+
+  config.define_derived_metadata(:file_path => /spec\/lib\/miq_automation_engine\/models/) do |metadata|
+    metadata[:type] ||= :model
+  end
+
+  config.include Spec::Support::AutomationHelper
```
Comment (bdunne) [inline spec/spec_helper.rb]: I don't think you need this

**Comment by Fryguy** [comment]
I've replaced this with #3 , fixed my issues above, and this can be closed in favor of #3

**Pull request closed without merging**